File: paimon-common/src/main/java/org/apache/paimon/predicate/VectorSearch.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.predicate;

import org.apache.paimon.utils.RoaringNavigableMap64;

import javax.annotation.Nullable;

import java.io.Serializable;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * VectorSearch to perform vector similarity search.
 *
 * <p>This is an internal pushdown representation. Use {@code Table.newVectorSearchBuilder()} to
 * configure vector search from Java.
 */
public class VectorSearch implements Serializable {

    private static final long serialVersionUID = 1L;

    private final float[] vector;
    private final String fieldName;
    private final int limit;
    private final Map<String, String> options;

    @Nullable private RoaringNavigableMap64 includeRowIds;

    public VectorSearch(float[] vector, int limit, String fieldName) {
        this(vector, limit, fieldName, Collections.emptyMap());
    }

    public VectorSearch(float[] vector, int limit, String fieldName, Map<String, String> options) {
        if (vector == null) {
            throw new IllegalArgumentException("Search vector cannot be null");
        }
        if (limit <= 0) {
            throw new IllegalArgumentException("Limit must be positive, got: " + limit);
        }
        if (fieldName == null || fieldName.isEmpty()) {
            throw new IllegalArgumentException("Field name cannot be null or empty");
        }
        this.vector = vector;
        this.limit = limit;
        this.fieldName = fieldName;
        this.options =
                options == null
                        ? Collections.emptyMap()
                        : Collections.unmodifiableMap(new HashMap<>(options));
    }

    public float[] vector() {
        return vector;
    }

    public int limit() {
        return limit;
    }

    public String fieldName() {
        return fieldName;
    }

    public Map<String, String> options() {
        return options == null ? Collections.emptyMap() : options;
    }

    public RoaringNavigableMap64 includeRowIds() {
        return includeRowIds;
    }

    public VectorSearch withIncludeRowIds(RoaringNavigableMap64 includeRowIds) {
        this.includeRowIds = includeRowIds;
        return this;
    }

    public VectorSearch offsetRange(long from, long to) {
        if (includeRowIds != null) {
            VectorSearch target = new VectorSearch(vector, limit, fieldName, options());
            target.withIncludeRowIds(VectorSearchUtils.offsetRowIds(includeRowIds, from, to));
            return target;
        }
        return this;
    }

    @Override
    public String toString() {
        return String.format("FieldName(%s), Limit(%s)", fieldName, limit);
    }
}

File: paimon-python/pypaimon/globalindex/vector_search.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

"""VectorSearch for performing vector similarity search."""

from dataclasses import dataclass, field
from typing import List, Optional, Union
import numpy as np


@dataclass
class VectorSearch:
    """
    VectorSearch to perform vector similarity search.
    
    Attributes:
        vector: The query vector (float[] or byte[])
        limit: Maximum number of results to return
        field_name: Name of the vector field to search
        include_row_ids: Optional bitmap of row IDs to include in search
    """

    vector: Union[List[float], np.ndarray]
    limit: int
    field_name: str
    include_row_ids: Optional['RoaringBitmap64'] = field(default=None)

    def __post_init__(self):
        if self.vector is None:
            raise ValueError("Search vector cannot be None")
        if self.limit <= 0:
            raise ValueError(f"Limit must be positive, got: {self.limit}")
        if not self.field_name:
            raise ValueError("Field name cannot be null or empty")
        
        # Convert list to numpy array if needed
        if isinstance(self.vector, list):
            self.vector = np.array(self.vector, dtype=np.float32)

    def with_include_row_ids(self, include_row_ids: 'RoaringBitmap64') -> 'VectorSearch':
        """Return a new VectorSearch with the specified include_row_ids."""
        return VectorSearch(
            vector=self.vector,
            limit=self.limit,
            field_name=self.field_name,
            include_row_ids=include_row_ids
        )

    def offset_range(self, from_: int, to: int) -> 'VectorSearch':
        """
        Create a new VectorSearch with include_row_ids offset to the given range.
        """
        from pypaimon.utils.roaring_bitmap import RoaringBitmap64

        if self.include_row_ids is not None:
            range_bitmap = RoaringBitmap64()
            range_bitmap.add_range(from_, to)
            and_result = RoaringBitmap64.and_(range_bitmap, self.include_row_ids)
            
            offset_bitmap = RoaringBitmap64()
            for row_id in and_result:
                offset_bitmap.add(row_id - from_)
            
            return VectorSearch(
                vector=self.vector,
                limit=self.limit,
                field_name=self.field_name,
                include_row_ids=offset_bitmap
            )
        return self

    def visit(self, visitor: 'GlobalIndexReader') -> Optional['GlobalIndexResult']:
        """Visit the global index reader with this vector search."""
        return visitor.visit_vector_search(self)

    def __repr__(self) -> str:
        return f"VectorSearch(field_name={self.field_name}, limit={self.limit})"
